p2p/enr: initial implementation

nonsense opened the pull request:
initial implementation of ENR according to https://github.com/ethereum/EIPs/pull/778

TODO:
- [x] compare RLP encoding/decoding with Python impl at https://github.com/fjl/p2p-drafts.
- [x] reject records longer than 300 bytes.
- [x] set `r.raw` on RLP decode, so that record is ready for RLP encoding after RLP decoding.
- [x] sort `pairs` on Set, use binary search on `Load`
- [x] ensure that setting the same key multiple times will update it instead of including it twice.
- [x] get rid of double-RLP encoding of `pair.value` - once in `Set()` and once is `signAndEncode()`
- [x] add documentation
- [x] use binary search on `Set`
- [x] add more tests to check decode/encode and sorting

nonsense commented on p2p/enr/enr.go:
```diff
@@ -0,0 +1,393 @@
+	return nil
+}
+
+func wrapList(c []byte) []byte {
```
This feels like a hack. I am looking to see if there is a nicer way to wrap a list.

nonsense commented on p2p/enr/enr.go:
```diff
@@ -0,0 +1,396 @@
+	}
+
+	digest := crypto.Keccak256Hash(content)
+	pubkey2, err := crypto.Ecrecover(digest.Bytes(), e.signature)
```
The signature returned from the python impl has a length of 64, and recovering of the public key fails.

fjl replied:
You can use `btcec.Signature.Verify` instead: https://godoc.org/github.com/btcsuite/btcd/btcec#example-package--VerifySignature

fjl commented on p2p/enr/enr.go:
```diff
@@ -0,0 +1,396 @@
+	return signature, raw, nil
+}
+
+func (e *ENR) SerialisedContent() ([]byte, error) {
```
No need to export this method.

fjl commented on p2p/enr/enr.go:
```diff
@@ -0,0 +1,396 @@
+		}
+	}
+
+	return wrapList(buffer.Bytes()), nil
```
You can avoid most of the RLP hackery by using higher-level types instead. To encode a record, put elements into `[]interface{}` then call 
`rlp.EncodeToBytes`. 

```go
list := []interface{}{nil, e.seq}
// add k/v elements
for _, r := range e.records {
    list = append(list, k, v)
}
// sign the tail of the list
sigcontent, err := rlp.EncodeToBytes(list[1:])
...
// add signature to front
list[0] = sig
record, err := rlp.EncodeToBytes(list)
```

fjl commented:
Something I forgot in the review summary earlier: the terms are slightly off in your implementation.
A "record" is the whole thing. key/value "pairs" are the elements of a "record". This should be reflected in identifiers.

nonsense commented:
@fjl thanks for the comments. The proposed API is indeed better. I think I addressed your comments, however I still haven't confirmed the RLP encoding/decoding vs the Python impl. - there might be some differences there. Will ping you when this is ready for review.

fjl commented on p2p/enr/discv5.go:
```diff
@@ -0,0 +1,41 @@
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+type DiscV5 uint32
```
UDP ports are uint16.

nonsense replied:
Done.

fjl commented on p2p/enr/discv5.go:
```diff
@@ -0,0 +1,41 @@
+		return err
+	}
+	return nil
+}
```
You can remove rlp encoder methods from Discv5 because the default encoding of uints does the same thing.

fjl replied:
Please rename `DiscV5` -> `DiscPort`.

nonsense replied:
Done.

fjl commented on p2p/enr/enr.go:
```diff
@@ -0,0 +1,293 @@
+}
+
+type pair struct {
+	k []byte
```
Package rlp supports strings. Please make `k` a string to avoid conversions.

nonsense replied:
Done.

fjl commented on p2p/enr/id.go:
```diff
@@ -0,0 +1,41 @@
+		return err
+	}
+	return nil
+}
```
You can remove rlp encoding methods from ID.

nonsense replied:
Done.

fjl commented on p2p/enr/secp256k1.go:
```diff
@@ -0,0 +1,41 @@
+		return err
+	}
+	return nil
+}
```
Please use 

```go
type Secp256k1 ecdsa.PublicKey
```

You already need to parse the public key for signature verification, might as well do it at decoding time.
The additional advantage will be that users can't get invalid curve points by accident.

fjl commented on p2p/enr/enr.go:
```diff
@@ -0,0 +1,293 @@
+
+	id := ID(ID_SECP256k1_KECCAK)
+
+	r.Set(id)
```
You can be less verbose and write `r.Set(ID(ISECP256k1_KECCAK))`.

nonsense replied:
Done.

fjl commented:
For your convenience, output of `make run` from p2p-drafts/testvec should be a lot more helpful now ;).

fjl commented:
I'll re-review when the rest of the boxes are checked ;)

nonsense commented:
@fjl great, will ping you when the TODO list is addressed.

nonsense commented:
@fjl @zsfelfoldi I think I addressed all comments and TODOs. This should be ready for review now.

zsfelfoldi commented on p2p/enr/keys.go:
```diff
@@ -0,0 +1,160 @@
+// To define a new key that is to be included in a node record,
+// create a Go type that satisfies this interface. The type should
+// also implement rlp.Decoder if additional checks are needed on the value.
+type Key interface {
```
This name confused me at first, I thought it is just the key from the key/value pair. Not sure what I'd like to suggest instead though, I would probably call it Entry or Item but I am not good at naming stuff :)

nonsense replied:
`Entry` is not a bad name for this interface. I can see how `Key` could confuse someone new to the package.

nonsense replied:
I just pushed a commit that renames `Key` to `Entry` where I found it to actually mean `Entry`, rather than `key` from a `key/value pair`. I am not 100% convinced the new version is better, probably because I am already used to the `Key` interface, but I don't have strong opinion on the name.

@fjl @zsfelfoldi let me know what you think, I am fine with both names to be honest.

zsfelfoldi commented on p2p/enr/enr.go:
```diff
@@ -0,0 +1,288 @@
+// Set adds or updates the given key in the record.
+// It panics if the value can't be encoded.
+func (r *Record) Set(k Key) {
+	r.signature = nil
```
Shouldn't this also invalidate r.raw? (probably you are only checking r.signature for nil but maybe it would be nicer to not leave an invalid r.raw there)

nonsense replied:
I agree - we should invalidate it - someone could try to decode the record, after having set a key and changed its contents. Pushed a changed.

zsfelfoldi commented on p2p/enr/keys.go:
```diff
@@ -0,0 +1,160 @@
+// DiscPort is the "discv5" key, which holds the UDP port for discovery v5.
+type DiscPort uint16
+
+func (v DiscPort) ENRKey() string { return "discv5" }
```
Why not just "port" or "udp"? Discovery v5-specific port will go away soon, the UDP port used by discovery will not depend on the version of the discovery protocol so we should not name this field after that.

nonsense replied:
I honestly don't know enough about the roadmap to comment on that. Depending on how `soon` is soon, and when exactly ENR will be integrated with the rest of the codebase, we can decide whether to keep this `DiscPort` entry type, or remove it and just add `Port`.

I guess we can postpone this decision until Discovery v5-specific port actually goes away?

nonsense commented:
@zsfelfoldi I think I addressed your comments. @fjl please review the `Key` rename to `Entry` and if you are not happy with it, feel free to just revert the commit. I think `Entry` reflects slightly better the semantics of the interface.

karalabe commented:
Please bump the copyright headers to 2017 :)

fjl commented:
The interface is called key because it assigns the key to a key/value pair.
